File: targets/python-flask/routes/xml.py
"""XML ingestion routes: XXE plus a defused variant."""
from flask import Blueprint, request, jsonify

from lxml import etree
from defusedxml import ElementTree as safe_et

xml_bp = Blueprint("xml", __name__, url_prefix="/xml")


# VULN: py-xxe-01 (xxe, cwe-611) [medium]
@xml_bp.route("/parse", methods=["POST"])
def parse_xml():
    data = request.get_data()
    parser = etree.XMLParser(resolve_entities=True, no_network=False)
    root = etree.fromstring(data, parser)
    return jsonify({"tag": root.tag, "text": root.text})


# SAFE: py-safe-03 (mimics xxe) - defusedxml rejects entities
@xml_bp.route("/parse_safe", methods=["POST"])
def parse_xml_safe():
    data = request.get_data()
    root = safe_et.fromstring(data)
    return jsonify({"tag": root.tag, "text": root.text})

File: targets/js-ts-express/routes/xml.ts
// XML ingestion routes.
import { Router, Request, Response } from 'express';

const libxmljs = require('libxmljs');

const router = Router();

router.post('/parse', (req: Request, res: Response) => {
  const xml = String(req.body);
  const doc = libxmljs.parseXml(xml, { noent: true, noclean: true });
  res.json({ root: doc.root().name(), text: doc.root().text() });
});

router.post('/parse_safe', (req: Request, res: Response) => {
  const xml = String(req.body);
  const doc = libxmljs.parseXml(xml, { noent: false });
  res.json({ root: doc.root().name(), text: doc.root().text() });
});

export default router;
